test/cpp/microbenchmarks/bm_binder_transport.cc

```diff
@@ -0,0 +1,515 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <benchmark/benchmark.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+#include <grpcpp/support/channel_arguments.h>
+#include <string.h>
+#include <memory>
+#include <queue>
+#include <sstream>
+#include "src/core/ext/transport/binder/transport/binder_transport.h"
+#include "src/core/lib/iomgr/closure.h"
+#include "src/core/lib/slice/slice_internal.h"
+#include "src/core/lib/transport/static_metadata.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/microbenchmarks/helpers.h"
+#include "test/cpp/util/test_config.h"
+
+////////////////////////////////////////////////////////////////////////////////
+// Helper classes
+//
+
+namespace grpc_binder {
+
+class PhonyReadableParcel : public ReadableParcel {
+ public:
+  PhonyReadableParcel() : ptr_(0) {}
+
+  template <class T>
+  void PushInput(T value) {
+    storage_.push_back(std::move(value));
+  }
+
+  absl::Status ReadInt32(int32_t* data) const override {
+    *data = absl::get<int32_t>(storage_[ptr_++]);
+    return absl::OkStatus();
+  }
+  absl::Status ReadBinder(std::unique_ptr<Binder>* data) const override;
+  absl::Status ReadByteArray(std::string* data) const override {
+    const auto& byte_array = absl::get<std::vector<int8_t>>(storage_[ptr_++]);
+    data->resize(byte_array.size());
+    for (size_t i = 0; i < data->size(); ++i) {
+      (*data)[i] = byte_array[i];
+    }
+    return absl::OkStatus();
+  }
+  absl::Status ReadString(char data[111]) const override {
+    const std::string& str = absl::get<std::string>(storage_[ptr_++]);
+    for (size_t i = 0; i < str.size(); ++i) {
+      data[i] = str[i];
+    }
+    return absl::OkStatus();
+  }
+
+ private:
+  std::vector<absl::variant<int, void*, std::vector<int8_t>, std::string>>
+      storage_;
+  mutable size_t ptr_;
```

Comment:
This is because the interface of `ReadableParcel` is marked as `const`. I forgot the reason why I did that in the first place (probably to match with the real Android binder since the NDK binder "read" methods all take a `const AParcel*`?). 

I can change it to a non-const interface and drop the `mutable` member if you think that's better.

Reply:
Yeah that would be better